### ISSUE TITLE
Deprecate existing Chado Query API in favour of new BioDB API

### ISSUE THREAD

**Issue by laceysanderson** [tripal/tripal#1341]
The new BioDB API extends the Drupal Database API to handle multiple PostgreSQL schema and cross-schema querying -something which is obviously needed for Tripal for ideal Chado integration. Specifically, this allows us to use the Drupal style of querying with Chado as shown in the following example:

```php
// Initiate the new Tripal BioDB service with Chado integration.
$biodb = \Drupal::service('tripal_chado.database');
// Start a select query on the feature table.
$query = $biodb->select('feature', 'x');
// Add JOINS to the organism and cvterm tables.
$query->join('organism', 'o', 'x.organism_id = o.organism_id');
$query->join('cvterm', 'cvt', 'x.type_id = cvt.cvterm_id');
// Exclude obsolete features.
$query->condition('x.is_obsolete', 'f', '=');
// Add the columns you want to be returned.
$query->fields('x', ['name', 'uniquename']);
$query->fields('o', ['genus', 'species', 'common_name']);
$query->addField('cvt', 'name', 'feature_type');
// Indicate the range of results you want.
$query->range(0, 10);
// Finally execute the above build query.
$result = $query->execute();
foreach ($result as $record) {
  // Do something with each record from the chado database.
}
```

Which would be the equivalent of the following Chado Query:

```sql
SELECT x.name, x.uniquename, cvt.name as feature_type, o.genus, o.species, o.common_name
FROM chado.feature x
LEFT JOIN chado.organism o ON x.organism_id = o.organism_id
LEFT JOIN chado.cvterm cvt ON x.type_id = cvt.cvterm_id
WHERE x.is_obsolete = f
LIMIT 10;
```

Tripal will be moving to this new Drupalized way of building queries and thus the old function-based Chado Query API will become deprecated. That said, it will remain around for multiple versions of Tripal 4.x to provide ample time for extension modules to upgrade to the new API.

Additionally, the insides of the old API will be replaced with this new Drupalized way of doing things in order to streamline maintenance of Tripal 4.x.

**Comment by spficklin** [comment]
Is there a way with BioDB to write straight SQL still?  I find there's always going to be some weird use case that we can't do with the Drupal classes and since BioDB uses that approach I'm just worried we may loose some flexibility when needed.

**Comment by spficklin** [comment]
Also, replacing the innards of the old Chado API functions will be a lot of work I suspect.   Given we are limited on time perhaps we just mark them as deprecated, leave them as is and then later remove them from a future release.

**Comment by laceysanderson** [comment]
> Is there a way with BioDB to write straight SQL still? I find there's always going to be some weird use case that we can't do with the Drupal classes and since BioDB uses that approach I'm just worried we may loose some flexibility when needed.

Yes you can still write straight SQL using the BioDB API so no lost functionality there! These code samples from @guignonv show that:

## Static query example on feature table in default Chado schema.

```php
$biodb = \Drupal::service('tripal_chado.database');
$sql_query = 'SELECT name, residues FROM {1:feature} x WHERE x.is_obsolete = :obsolete LIMIT 0, 10;';
$results = $biodb->query($sql_query, [':obsolete' => 'f']);
foreach ($results as $record) {
  echo $record->name . "\n";
}
```

See also [Drupal static queries](https://www.drupal.org/docs/drupal-apis/database-api/static-queries) and [result sets](https://www.drupal.org/docs/8/api/database-api/result-sets).

## Cross schema queries.

```php
$biodb = \Drupal::service('tripal_chado.database');
$biodb->setSchemaName('chado1');
$biodb->addExtraSchema('chado2');
$sql = "
  SELECT * FROM
    {1:feature} f1,
    {2:feature} f2,
    {node_field_data} fd
  WHERE fd.title = f1.uniquename
    AND f1.uniquename = f2.uniquename;";
$results = $biodb->query($sql);
```

**Comment by laceysanderson** [comment]
> Also, replacing the innards of the old Chado API functions will be a lot of work I suspect. Given we are limited on time perhaps we just mark them as deprecated, leave them as is and then later remove them from a future release.

Mmmm.... this is likely true. I thought this endeavour might be a good way for me to learn the ins and outs of the new API but it may end up not being worth it... I'll make sure to dial my idealism back a bit here ;-p

**Comment by spficklin** [comment]
One other benefit for leaving them is there's no accidental, inadvertent changing of behavior in them from the rewrite.

**Comment by laceysanderson** [comment]
This issue is remaining open to let people know of this deprecation and to remind us to add the deprecation notices to the code.

**Comment by laceysanderson** [comment]
@pdtouch this is the issue that your work on deprecations will be in reference to. You can use the branch name `tv4g2-issue1341-chadoQueryDeprecations` in your fork.

**Comment by laceysanderson** [comment]
We need to actually finish deprecating these 🙈 I have started updating the change record for this in https://github.com/tripal/tripal_doc/pull/109
